### ISSUE TITLE
Assertion in CDT_plus_2

### ISSUE THREAD

**Issue by VincentBoumard** [CGAL/cgal#5055]
## Issue Details

Hello, I get the following assert : 

assertion at D:\Projects\TC2\Main\WorkY\SystemLibs\PC64\CGAL-5.0.2\include\CGAL/Triangulation_2/internal/Polyline_constraint_hierarchy_2.h line 1016: 
 (false).

using this sample :

## Source Code

```c
#include <CGAL/Constrained_triangulation_plus_2.h>
#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
#include <CGAL/Projection_traits_xy_3.h>
#include <CGAL/Triangulation_face_base_with_info_2.h>
#include <CGAL/Triangulation_vertex_base_with_id_2.h>

	typedef CGAL::Epick Kernel;
    typedef Kernel::FT FieldNumberType;
    typedef Kernel::Point_2 Point2;
    typedef Kernel::Point_3 Point3;
	struct FaceInfo2
    {
	    unsigned long long m_id;
    };
	typedef CGAL::Projection_traits_xy_3<Kernel> TriangulationTraits;
    typedef CGAL::Triangulation_vertex_base_with_id_2<TriangulationTraits> VertexBaseWithId;
    typedef CGAL::Triangulation_vertex_base_2<TriangulationTraits, VertexBaseWithId> VertexBase;
    typedef CGAL::Triangulation_face_base_with_info_2<FaceInfo2, Kernel> FaceBaseWithInfo;
    typedef CGAL::Constrained_triangulation_face_base_2<TriangulationTraits, FaceBaseWithInfo> FaceBase;
    typedef CGAL::Triangulation_data_structure_2<VertexBase, FaceBase> TriangulationData;
    typedef CGAL::Constrained_Delaunay_triangulation_2<TriangulationTraits, TriangulationData, CGAL::Exact_predicates_tag> ConstrainedTriangulation;
    typedef CGAL::Constrained_triangulation_plus_2<ConstrainedTriangulation> CDT;
	
	void TriangulateCrash()
    {
		Point3 suspects[4][4] =
		{
			{
				Point3(4.871738281250000e+02,8.464453125000000e+01, 0),
				Point3(4.868144531250000e+02,7.965625000000000e+01, 0),
				Point3(4.893066406250000e+02,7.947656250000000e+01, 0),
				Point3(4.896660156250000e+02,8.446484375000000e+01, 0)
			},

			{
				Point3(4.869394531250000e+02,8.163281250000000e+01, 0),
				Point3(4.866582031250000e+02,7.664062500000000e+01, 0),
				Point3(4.891542968750000e+02,7.650000000000000e+01, 0),
				Point3(4.894355468750000e+02,8.149218750000000e+01, 0)
			},

			{
				Point3(4.869394531250000e+02,8.163281250000000e+01, 0),
				Point3(4.866582031250000e+02,7.664062500000000e+01, 0),
				Point3(4.891542968750000e+02,7.650000000000000e+01, 0),
				Point3(4.894355468750000e+02,8.149218750000000e+01, 0)
			},

			{
				Point3(4.871738281250000e+02,8.464453125000000e+01, 0),
				Point3(4.868144531250000e+02,7.965625000000000e+01, 0),
				Point3(4.893066406250000e+02,7.947656250000000e+01, 0),
				Point3(4.896660156250000e+02,8.446484375000000e+01, 0)
			}

		};

		CDT	cdt;
		for (int i = 0; i < 4; i++)
		{
			cdt.insert_constraint(suspects[i][0], suspects[i][1]);
			cdt.insert_constraint(suspects[i][1], suspects[i][2]);
			cdt.insert_constraint(suspects[i][2], suspects[i][3]);
			cdt.insert_constraint(suspects[i][3], suspects[i][0]);
			cdt.insert_constraint(suspects[i][3], suspects[i][1]);
			cdt.insert_constraint(suspects[i][2], suspects[i][0]);
		}
    }
```

## Environment

* Operating system : Windows 64 bits
* Compiler: VS 2015
* Release or debug mode: both
* CGAL version: 5.0.2

Thank you !

**Comment by lrineau** [comment]
I confirm the bug, I can reduce it to:
```
cdt.insert_constraint(Point3(489.306640625, 79.4765625, 0), Point3(489.666015625, 84.46484375, 0));
cdt.insert_constraint(Point3(489.154296875, 76.5, 0), Point3(489.435546875, 81.4921875, 0));
cdt.insert_constraint(Point3(489.154296875, 76.5, 0), Point3(489.435546875, 81.4921875, 0));
cdt.insert_constraint(Point3(489.306640625, 79.4765625, 0), Point3(489.666015625, 84.46484375, 0));
```
that is:
- insert AB
- insert CD (that intersects AB)
- insert CD
- insert AB

**Comment by lrineau** [comment]
@afabri The CDT_plus_2 should avoid inserting twice the same sub-constraint. That is the bug.

**Comment by lrineau** [comment]
Before CGAL-4.7, a constraint was a `pair<Vertex_handle, Vertex_handle>`. The constraints hierarchy had a map constraint-to-subconstraint, and it allowed to check if a constraint was already inserted in the triangulation. Since CGAL-4.7, a constraint is a polyline, and that check has been removed because we have no efficient way to check for the equality of two polylines.

Anyway, even at the time of CGAL-4.6, the CDT_2 was very fragile when a user tried to insert AB, then CD intersecting it, then BA (AB in the opposite direction).

@VincentBoumard Do you have a way to avoid the insertion of twice the same constraint, with checks before the calls to the CDT?

**Comment by lrineau** [comment]
@VincentBoumard Alternatively, you can define the macro `CGAL_CDT_2_INTERSECTION_SNAPPING_ULP_DISTANCE` to `32`, and it will work: during the construction of intersection points, the newly constructed points will be snapped onto close points, if their coordinates are similar enough (up to 5 bits of precision, or 32 [ULP](https://en.wikipedia.org/wiki/Unit_in_the_last_place), among the 53 bits available in doubles).

**Comment by VincentBoumard** [comment]
Hey @lrineau , using the define (even when upping the imprecision) is not enough to remove the crash. The solution will have to be on my end by not inserting a constraint that's already added. Could you confirm that once a contrain AB is inserted, subsequent adds of both AB and BA should be avoided ?
Thank you !

**Comment by lrineau** [comment]
> Could you confirm that once a contrain AB is inserted, subsequent adds of both AB and BA should be avoided ?

Yes, exactly: because that segment AB can be decomposed into almost-collinear pieces, by other intersecting constraints. And if you insert AB or BA again, the situation becomes very degenerate: almost- (but non-) collinear segments intersecting other segments.

**Comment by VincentBoumard** [comment]
Hello @lrineau, I have managed to isolate a sample not containing any duplicate constraints. The error occurs when inserting the last constraint, which is similar to but not equal to one already inserted. It is also different from all sub constraints in the triangulation.

```c++
	CDT	cdt;
		cdt.insert_constraint(Point3(7.22060172459952354984e+02, 5.46431441157868448499e+02, 1.09259033203125000000e+02), Point3(7.21605164369421572701e+02, 5.47286947218267187054e+02, 1.09231933593750000000e+02));
		cdt.insert_constraint(Point3(4.89054338074482529919e+02, 7.34795619763430352123e+01, 1.19170654296875000000e+02), Point3(4.88992971641138979066e+02, 7.06742393092105203323e+01, 1.19197509765625000000e+02));
		cdt.insert_constraint(Point3(6.58294921875000000000e+02, 4.94029296875000000000e+02, 1.08803459167480468750e+02), Point3(6.55601638506381050320e+02, 4.92704923416975930195e+02, 1.08774169921875000000e+02));
		cdt.insert_constraint(Point3(7.48077080682167661507e+02, 5.16511846264136920581e+02, 1.09748291015625000000e+02), Point3(7.46966796875000000000e+02, 5.16371093750000000000e+02, 1.09758056640625000000e+02));
		cdt.insert_constraint(Point3(5.59349064841414701732e+02, 4.42600717751872252848e+02, 1.08587158203125000000e+02), Point3(5.56660156250000000000e+02, 4.41277343750000000000e+02, 1.08660644531250000000e+02));
		cdt.insert_constraint(Point3(6.36340512288503987293e+02, 4.83223199100345823354e+02, 1.08523681640625000000e+02), Point3(6.36339843750000000000e+02, 4.83224609375000000000e+02, 1.08523681640625000000e+02));
		cdt.insert_constraint(Point3(5.72431981051620141443e+02, 4.62372208487347847949e+02, 1.08266105651855468750e+02), Point3(5.71998046875000000000e+02, 4.61550781250000000000e+02, 1.08266105651855468750e+02));
		cdt.insert_constraint(Point3(5.54615246238433428516e+02, 3.99093747119970316817e+02, 1.08986572265625000000e+02), Point3(5.53910156250000000000e+02, 3.96179687500000000000e+02, 1.09051513671875000000e+02));
		cdt.insert_constraint(Point3(6.29072265625000000000e+02, 4.77775390625000000000e+02, 1.08448486328125000000e+02), Point3(6.29071750798135099103e+02, 4.77776482510548703431e+02, 1.08448486328125000000e+02));
		cdt.insert_constraint(Point3(4.89089689771187522638e+02, 7.46450533344111875067e+01, 1.19143554687500000000e+02), Point3(4.89006004028320262478e+02, 7.14945312499999943157e+01, 1.19197509765625000000e+02));
		cdt.insert_constraint(Point3(4.89054338074482529919e+02, 7.34795619763430352123e+01, 1.19170654296875000000e+02), Point3(4.88992951403166102864e+02, 7.06733141447368495847e+01, 1.19197509765625000000e+02));
```

Thank you !

**Comment by VincentBoumard** [comment]
Hello @lrineau , do you have an update on this problem ? Thanks !

**Comment by afabri** [comment]
I've put [`issue_5055.cpp` ](https://gist.github.com/afabri/cfdf5d5929fa42175112d483f6932c4e) on gist.github.com,  I cannot reproduce the error (an assertion) with VC2019.

**Comment by lrineau** [comment]
> I've put [`issue_5055.cpp` ](https://gist.github.com/afabri/cfdf5d5929fa42175112d483f6932c4e) on gist.github.com, I cannot reproduce the error (an assertion) with VC2019.

I cannot reproduce either with clang version `10.0.1 (Fedora 10.0.1-3.fc32)` on Linux. I have tried with cgal `master` (currently a60a9e638dd, identical to `v5.2-beta1`), and with CGAL-5.0.4.

**Comment by VincentBoumard** [comment]
I've confirmed that the crash still happens with the sample I provided on CGAL version 5.1 (using VS 2019 compiler). I'll have to test version 5.2 or 5.0.4 which I don't have currently

**Comment by lrineau** [comment]
> > I've put [`issue_5055.cpp` ](https://gist.github.com/afabri/cfdf5d5929fa42175112d483f6932c4e) on gist.github.com, I cannot reproduce the error (an assertion) with VC2019.
> 
> I cannot reproduce either with clang version `10.0.1 (Fedora 10.0.1-3.fc32)` on Linux. I have tried with cgal `master` (currently [a60a9e6](https://github.com/CGAL/cgal/commit/a60a9e638dd243e68f8efa19dc73b3b0adf83ce0), identical to `v5.2-beta1`), and with CGAL-5.0.4.

I have tried as well with CGAL-5.1.1 and g++ version 10.2.1, still on Linux Fedora 32, and it worked.

Maybe something specific to Windows, or MSVC.

**Comment by hanilan** [comment]
Hello there, 
I'm interesting in this issue as I had similar crashes with CDT (hcl was nullptr). I took the sample just above and I had the crash too, but if I change the floating point model from precise to fast (https://software.intel.com/content/www/us/en/develop/documentation/cpp-compiler-developer-guide-and-reference/top/compiler-reference/floating-point-operations/understanding-floating-point-operations/using-the-fp-model-fp-option.html), I did not reproduce it. I'm using Visuall C++ 2019 version 16.7.7 with cl and CGAL 5.1. Maybe it can explain why you didn't trigger the crash.
